### PR TITLE
NIAD-1294: missing message header resource

### PR DESCRIPTION
**Pull request by bartek-sarul**
## Description

Please explain the changes you have made here.

## Jira Ticket

Please add a link to the related Jira ticket (if any).

## Checklist

These are items (excluding GitHub Checks) which should be confirmed before a branch is ready to merge.

- [ ] Acceptance Criteria met
- [ ] Commit messages are meaningful
- [ ] Manually tested
- [ ] Self-reviewed your code
- [ ] Code reviewed from two other developers
- [ ] If your pull request depends on any other, please link them in the description
- [ ] main branch is passing/green on CI
- [ ] Add/update any relevant documentation

### CODE REVIEW

File: src/intTest/resources/fhir/multi_pathology_msg1.fhir.json [hunk 1]
```diff
@@ -1,23 +1,88 @@
...
+                                "url": "LocalExtension",
+                                "valueString": "None"
+                            }
+                        ]
+                    }
+                ],
+                "event": {
+                    "system": "https://fhir.nhs.uk/STU3/CodeSystem/ITK-MessageEvent-2",
+                    "code": "ITK012M",
+                    "display": "ITK National Pathology"
```
Comment (bartek-sarul) [inline src/intTest/resources/fhir/multi_pathology_msg1.fhir.json:60]: There is no code for Screening

File: src/intTest/resources/fhir/multi_pathology_msg1.fhir.json [hunk 2]
```diff
@@ -89,10 +154,22 @@
             }
         },
         {
-            "fullUrl": "urn:uuid:b019bdf7-53fe-42a6-89f4-24a6ab5be2c4",
+            "fullUrl": "urn:uuid:df8c2de0-c846-4041-a2e9-5cdf1e6390aa",
+            "resource": {
+                "resourceType": "Practitioner",
```
Comment (bartek-sarul) [inline src/intTest/resources/fhir/multi_pathology_msg1.fhir.json:159]: So now we create a requester or performer Practitioner resource even if there is no information in EDIFACT

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/mesh/RecipientMailboxIdMappingsTest.java [hunk 3]
```diff
@@ -21,35 +19,14 @@ void setUp() {
...
-
-    @Test
-    void testGetRecipientMailboxIdForMessageNoRecipientToMailboxMappingsThrowsException() {
-
-        recipientMailboxIdMappings = new RecipientMailboxIdMappings("");
-
-        final OutboundMeshMessage message = new MeshMessage();
-
-        final MeshRecipientUnknownException exception = assertThrows(MeshRecipientUnknownException.class,
-            () -> recipientMailboxIdMappings.getRecipientMailboxId(message));
```
Comment (n-r-b) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/mesh/RecipientMailboxIdMappingsTest.java:49]: This test scenario could still be valid if an empty string was passed as the recipient

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class MessageHeaderMapper {
+    private final UUIDGenerator uuidGenerator;
+    private final ResourceFullUrlGenerator fullUrlGenerator;
+    private final RecipientMailboxIdMappings recipientMailboxIdMappings;
+
+    public MessageHeader mapToMessageHeader(
+            Message message, DiagnosticReport diagnosticReport,
+            Organization performingOrganization, Organization requestingOrganization) {
+        final var messageHeader = new MessageHeader();
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java:30]: Do we need to set the timestamp? I believe this is mandatory according to the MessageHeader profile

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        messageHeader.getMeta().addProfile(FhirProfiles.MESSAGE_HEADER);
+        messageHeader.setId(uuidGenerator.generateUUID());
+
+        messageHeader.addFocus(new Reference(fullUrlGenerator.generate(diagnosticReport)));
+
+        messageHeader.setResponsible(new Reference(fullUrlGenerator.generate(performingOrganization)));
+
+        var extension = messageHeader.addExtension();
+        extension.setUrl("https://fhir.nhs.uk/STU3/StructureDefinition/Extension-ITK-MessageHandling-2");
+        extension.addExtension(new Extension().setUrl("BusAckRequested").setValue(new BooleanType(true)));
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java:40]: Can be simplified slightly I think
```suggestion
        extension.addExtension().setUrl("BusAckRequested").setValue(new BooleanType(true));
```

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.stereotype.Component;
+import uk.nhs.digital.nhsconnect.lab.results.mesh.RecipientMailboxIdMappings;
+import uk.nhs.digital.nhsconnect.lab.results.model.FhirProfiles;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Message;
+import uk.nhs.digital.nhsconnect.lab.results.utils.ResourceFullUrlGenerator;
+import uk.nhs.digital.nhsconnect.lab.results.utils.UUIDGenerator;
+
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class MessageHeaderMapper {
```
Comment (n-r-b) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MessageHeaderMapper.java:22]: I don't see any unit tests for this class

File: src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/BundleMapper.java [hunk 7]
```diff
@@ -30,18 +30,22 @@ public Bundle mapToBundle(final MedicalReport medicalReport) {
...
+                medicalReport.getRequestingOrganization(),
+                medicalReport.getPerformingPractitioner(),
+                medicalReport.getPerformingOrganization(),
+                medicalReport.getTestRequestSummary(),
+                medicalReport.getTestReport()
+            ),
+            medicalReport.getSpecimens().stream(),
+            medicalReport.getTestResults().stream()
+        )
+            .flatMap(i -> i)
```
Comment (andrewwa-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/BundleMapper.java:47]: Interesting way to do it. A couple of alternatives that _might_ make it more obvious what's going on:
```suggestion
            .flatMap(Function.identity())
```
or
```suggestion
            .reduce(Stream.empty(), Stream::concat)
```

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MedicalReportMapperTest.java [hunk 8]
```diff
@@ -58,144 +59,60 @@
...
+        assertThat(medicalReport.getMessageHeader()).isEqualTo(messageHeader);
+        assertThat(medicalReport.getPatient()).isEqualTo(patient);
+        assertThat(medicalReport.getPerformingOrganization()).isEqualTo(performingOrganization);
+        assertThat(medicalReport.getPerformingPractitioner()).isEqualTo(performingPractitioner);
+        assertThat(medicalReport.getRequestingOrganization()).isEqualTo(requestingOrganization);
+        assertThat(medicalReport.getRequestingPractitioner()).isEqualTo(requestingPractitioner);
         assertThat(medicalReport.getTestReport()).isEqualTo(diagnosticReport);
+        assertThat(medicalReport.getTestRequestSummary()).isEqualTo(procedureRequest);
+        assertThat(medicalReport.getSpecimens()).isEqualTo(specimens);
+        assertThat(medicalReport.getTestResults()).isEqualTo(observations);
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/MedicalReportMapperTest.java:116]: Might be worth grouping these in an `assertAll()`

File: src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PractitionerMapperTest.java [hunk 9]
```diff
@@ -35,7 +34,14 @@
...
         when(message.getInvolvedParties()).thenReturn(Collections.emptyList());
 
-        assertThat(mapper.mapToRequestingPractitioner(message)).isEmpty();
+        Practitioner requestingPractitioner = mapper.mapToRequestingPractitioner(message);
+
+        assertAll(
+            () -> assertThat(requestingPractitioner.getName())
+                .hasSize(0),
+            () -> assertThat(requestingPractitioner.getIdentifier())
+                .hasSize(0)
```
Comment (andrewwa-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/PractitionerMapperTest.java:43]: `isEmpty` might be clearer
```suggestion
            () -> assertThat(requestingPractitioner.getName())
                .isEmpty(),
            () -> assertThat(requestingPractitioner.getIdentifier())
                .isEmpty()
```

**Review by andrewwa-kainos — APPROVED** [review]

**Review by n-r-b — APPROVED** [review]

**Pull request merged**